README
RLP 1.5 - solver for linear programming optimization problems
Copyright (C) 2006, 2009, 2015 Rui Carlos Gonçalves


For info about instalation, see file 'INSTALL'.
For info about program usage, see folder 'doc'.


This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.


ChangeLog
2015-08-20 Rui Carlos Gonçalves <rcgoncalves.pt@gmail.com>
	* Version 1.5
	* Updated LibRCG to the latest version.
	* Documentation translated to English.
	* Minor improvements to avoid compilation warnings.

2009-02-07 Rui Carlos Gonçalves <rcgoncalves.pt@gmail.com>
	* Version 1.1.
	* Updated LibRCG to latest version.
	* Documentation reviewed.
	* Program output translated to English.
	* src/main.c (opt): Added option to show program version.
	* src/rlp.c (simplex): Added test to check when the output file is
	NULL, so that the program does not try to print the matrix in such
	case.

2006-12-01 Rui Carlos Gonçalves <rcgoncalves.pt@gmail.com>
	* Version 1.0.


src/main.h
/**
 * Definition of functions used by main.
 *
 * @author Rui Carlos Gonçalves
 * @file main.h
 * @version 1.5
 * @date 08/2015
 */
#ifndef _MAIN_
#define _MAIN_

#include "array.h"
#include "read.h"

/**
 * From the info collected by the parser, creates a matrix representing the
 * problem to solve, according to the format required by function
 * <tt>simplex</tt>.
 *
 * @param prob  info about the problem.
 * @param varc  number of variables.
 * @param condc number of conditions.
 * @param eqc   number of conditions with an equality operator.
 * @param type  determines whether we have a maximization or minimization
 * problem (max.: -1, min.: 1).
 *
 * @return the matrix created.
 */
double* loadMatrix(Prob prob,int varc,int condc,int eqc,int type);

/**
 * Prints the result.
 *
 * @param vars  name of the variables of each position of the table.
 * @param tab   table resulting from applying the <em>Simplex</em> algorithm.
 * @param varob variable to maximize/minimize.
 * @param varc  number of variables.
 * @param condc number of conditions
 * @param type  determines whether we have a maximization or minimization
 * problem (max.: -1, min.: 1).
 */
void printRes(Array vars,double* tab,const char* varob,int varc,int condc,int type);

/**
 * Checks the options specified by the user.
 *
 * @param argc number of options.
 * @param argv value of the options.
 * @param in   address for the input option.
 * @param out  address for the output option.
 *
 * @return 1 if the <tt>tables</tt> option was set; 0 otherwise.
 */
int opt(int argc,char** argv,char** in,char** out);

#endif


src/read.c
/**
 * Definitions of data types and functions required to read and store the
 * data about the problem.
 *
 * @author Rui Carlos Gonçalves
 * @file read.c
 * @version 1.5
 * @date 08/2015
 */
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "read.h"

/**
 * Given a row (<tt>R</tt>), a column (<tt>C</tt>), and the number of columns
 * (<tt>NC</tt>) of a matrix, computes an equivalent 1D position.
 */
#define POS(R,C,NC) ((R)*(NC)+(C))

/// Error messages.
static char* errors[]={"ERROR! Function \'setCoefHead\' -> \'malloc\'.\n"
                      ,"ERROR! Function \'setCoefHead\' -> \'listInsertLst\'.\n"
                      ,"ERROR! Function \'setCoefHead\' -> \'arrayInsert\'.\n"
                      ,"ERROR! Function \'setCoefHead\' -> \'newExp\'.\n"
                      ,"ERROR! Function \'setCoefHead\' -> \'hashInsert\'.\n"
                      ,"ERROR! Function \'setCoefTail\' -> \'malloc\'.\n"
                      ,"ERROR! Function \'setCoefTail\' -> \'hashInsert\' or"
                      "\'arrayInsert\'.\n"
                      ,"ERROR! Function \'setCoefTail\' -> \'arrayInsert\'.\n"
                      };

//##############################################################################

Exp newExp(int size)
{
  Exp res;

  res=malloc(sizeof(SExp));

  if(res)
  {
    res->coefs=newArray(size);
    res->op=0;
    res->rhs=0;
  }

  return res;
}

//##############################################################################

Prob newProb()
{
  Prob res;

  res=malloc(sizeof(SProb));

  if(res)
  {
    res->pos=newHash(32,0.6,(int(*)(void*))varhash,(int(*)(void*,void*))strcmp);
    res->invpos=newArray(10);
    res->exps=newList();
  }

  return res;
}

//##############################################################################

int varhash(const char* varid)
{
  int i,res;

  for(i=0,res=0;varid[i];i++)
    res+=varid[i];

  return res;
}

//##############################################################################

int addCoefHead(Prob prob,const char* varid,int varc,double coef)
{
  int* vpos,pos,erro,res;
  double* tmp;
  Exp exp;

  vpos=&pos;
  res=0;

  if(hashGet(prob->pos,(void*)varid,(void**)&vpos))
  {
    vpos=malloc(sizeof(int));

    if(!vpos){
      fputs(errors[0],stderr);
      exit(1);
    }

    *vpos=varc;
    erro=hashInsert(prob->pos,(void**)varid,vpos,0);
    erro+=arrayInsert(prob->invpos,varc,(void**)varid,0);

    if(erro){
      fputs(errors[4],stderr);
      exit(5);
    }

    res=1;
  }

  exp=newExp((varc+1>10)?(varc+1):10);

  if(!exp){
    fputs(errors[3],stderr);
    exit(4);
  }

  erro=listInsertLst(prob->exps,exp);

  if(erro){
    fputs(errors[1],stderr);
    exit(2);
  }

  tmp=malloc(sizeof(double));

  if(!tmp){
    fputs(errors[0],stderr);
    exit(1);
  }
    
  *tmp=coef;
  erro=arrayInsert(exp->coefs,*vpos,tmp,0);

  if(erro){
    fputs(errors[2],stderr);
    exit(3);
  }

  return res;
}

//##############################################################################

int addCoefTail(Prob prob,const char* varid,int varc,double coef)
{
  int* vpos,pos,res,error;
  double* tmp;
  Exp exp;

  vpos=&pos;
  res=0;

  if(hashGet(prob->pos,(void*)varid,(void**)&vpos))
  {
    vpos=malloc(sizeof(int));

    if(!vpos){
      fputs(errors[5],stderr);
      exit(6);
    }

    *vpos=varc;
    error=hashInsert(prob->pos,(void**)varid,vpos,0);
    error+=arrayInsert(prob->invpos,varc,(void**)varid,0);

    if(error)
    {
      fputs(errors[6],stderr);
      exit(7);
    }

    res=1;
  }

  listLst(prob->exps,(void**)&exp);
  tmp=malloc(sizeof(double));

  if(!tmp){
    fputs(errors[5],stderr);
    exit(6);
  }

  *tmp=coef;
  error=arrayInsert(exp->coefs,*vpos,tmp,0);

  if(error)
  {
    fputs(errors[7],stderr);
    fprintf(stderr,"\n%d\n\n",error);
    exit(8);
  }
 
  return res;
}

//##############################################################################

int setOpRHS(Prob prob,int op,double rhs)
{
  Exp exp;

  listLst(prob->exps,(void**)&exp);

  exp->op=op;
  exp->rhs=rhs;

  return 0;
}


src/read.h
/**
 * Definitions of data types and functions required to read and store the
 * data about the problem.
 *
 * @author Rui Carlos Gonçalves
 * @file read.h
 * @version 1.5
 * @date 08/2015
 */
#ifndef _READ_
#define _READ_

#include "array.h"
#include "hashmap.h"
#include "list.h"

/**
 * Expression structure.
 */
typedef struct sExp
{
  /// Coefficients of the variables.
  Array coefs;
  /// Operator of the expression.
  int op;
  /// Right-hand side of the expression.
  double rhs;
}SExp;

/**
 * Expression definition.
 */
typedef SExp* Exp;

/**
 * Problem structure.
 */
typedef struct sProb
{
  /// Variables and their indexes.
  HashMap pos;
  /// Variables on each index.
  Array invpos;
  /// List of expressions (objective function, and conditions). 
  List exps;
}SProb;

/**
 * Problem definition.
 */
typedef SProb* Prob;

//##############################################################################

/**
 * Creates an expression.
 *
 * @param arraysize size of the coefficients array.
 *
 * @return <tt>NULL</tt> if an error occurred; the new expression otherwise.
 */
Exp newExp(int arraysize);

/**
 * Creates a problem.
 *
 * @return <tt>NULL</tt> if an error occurred; the new problem otherwise.
 */
Prob newProb();

/**
 * Hash function for variables names.
 *
 * @param varid the variable identifier.
 *
 * @return the hash code of the variable.
 */
int varhash(const char* varid);

/**
 * Adds a new condition to the problem, and adds the coefficients of a variable.
 * If the variable did not exist in the variables set, it is added.
 *
 * @param prob  the problem.
 * @param varid the variable identifier.
 * @param varc  the index of the variable if it does not exist.
 * @param coef  the coefficient of the variable.
 *
 * @return 0 if the variable already existed; 0 otherwise.
 */
int addCoefHead(Prob prob,const char* varid,int varc,double coef);

/**
 * Adds the coefficient of a variable.
 * If the variable did not exist in the variables set, it is added.
 *
 * @param prob  the problem.
 * @param varid the variable identifier.
 * @param varc  the index of the variable if it does not exist.
 * @param coef  the coefficient of the variable.
 *
 * @return 0 if the variable already existed; 0 otherwise.
 */
int addCoefTail(Prob prob,const char* varid,int varc,double coef);

/**
 * Adds the relational operator and the right-hand side to a condition.
 *
 * @param prob the problem.
 * @param op   the identifier of the relational operator.
 * @param rhs  the right-hand side of the condition.
 *
 * @return 0.
 */
int setOpRHS(Prob prob,int op,double rhs);

#endif


src/main.c
/**
 * Definition of functions used by main.
 *
 * @author Rui Carlos Gonçalves
 * @file main.c
 * @version 1.5
 * @date 08/2015
 */
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include "main.h"

/**
 * Given a row (<tt>R</tt>), a column (<tt>C</tt>), and the number of columns
 * (<tt>NC</tt>) of a matrix, computes an equivalent 1D position.
 */
#define POS(R,C,NC) ((R)*(NC)+(C))

/// Error messages.
static char* errors[]={"ERROR! Function \'loadMatrix\' -> \'calloc\'.\n"
                      ,"ERROR! Function \'printRes\' -> \'newArray\'.\n"
                      ,"ERROR! Function \'printRes\' -> \'malloc\'.\n"
                      ,"ERROR! Function \'printRes\' -> \'arrayInsert\'.\n"
                      ,"ERROR! Invalid parameters.\n"
                      };

//##############################################################################

double* loadMatrix(Prob prob,int varc,int condc,int eqc,int type)
{
  double* res,*tmp;
  int row,col,k;
  Exp exp;

  k=varc+condc+eqc+2;
  res=calloc((condc+1+eqc)*(varc+condc+eqc+2),sizeof(double));

  if(!res){
    fputs(errors[0],stderr);
    exit(1);
  }

  listRemoveFst(prob->exps,(void**)&exp);

  res[POS(0,k-2,k)]=0;
  res[POS(0,k-1,k)]=0;

  for(col=0;col<k-2;col++)
  {
    arrayAt(exp->coefs,col,(void**)&tmp);
    res[POS(0,col,k)]=tmp?type*(*tmp):0;
  }

  arrayMap(exp->coefs,free);
  arrayDelete(exp->coefs);
  free(exp);

  for(row=1;!listRemoveFst(prob->exps,(void**)&exp);)
  {
    if(exp->op>-1)
    {
      res[POS(row,k-2,k)]=exp->rhs;
      res[POS(row,k-1,k)]=row+varc-1;

      for(col=0;col<varc;col++)
      {
        arrayAt(exp->coefs,col,(void**)&tmp);
        res[POS(row,col,k)]=tmp?*tmp:0;
      }

      row++;

      arrayMap(exp->coefs,free);
      arrayDelete(exp->coefs);
      free(exp);
    }

    if(exp->op<1)
    {
      res[POS(row,k-2,k)]=-1*(exp->rhs);
      res[POS(row,k-1,k)]=row+varc-1;

      for(col=0;col<varc;col++)
      {
        arrayAt(exp->coefs,col,(void**)&tmp);
        res[POS(row,col,k)]=tmp?-(*tmp):0;
      }

      row++;
    }
  }

  for(row=0;row<condc;row++)
    for(col=0;col<condc;col++)
      res[POS(row+1,col+varc,k)]=row==col;

  return res;
}

//##############################################################################

void printRes(Array vars,double* tab,const char* varob,int varc,int condc,int type)
{
  int i,k,erro;
  double* tmp;
  char* str;
  Array res;
  
  k=varc+condc+2;
  res=newArray(varc);

  if(!res){
    fputs(errors[1],stderr);
    exit(2);
  }

  for(i=1;i<condc+1;i++)
    if(tab[POS(i,k-1,k)]<varc)
    {
      tmp=malloc(sizeof(double));

      if(!tmp){
	fputs(errors[2],stderr);
        exit(3);
      }

      *tmp=tab[POS(i,k-2,k)];
      erro=arrayInsert(res,tab[POS(i,k-1,k)],tmp,0);

      if(erro){
	fputs(errors[3],stderr);
        exit(4);
      }
    }

  puts("====================");
  printf("%s = %f\n",varob,-type*tab[POS(0,k-2,k)]);
  puts("--------------------");
  for(i=0;i<varc;i++)
  {
    arrayAt(vars,i,(void**)&str);
    printf("%s =",str);
    if(!arrayAt(res,i,(void**)&tmp)) printf(" %f\n",*tmp);
    else printf(" 0.0\n");
  }
  puts("====================");

  arrayMap(res,free);
  arrayDelete(res);
}

//##############################################################################

int opt(int argc,char** argv,char** in,char** out)
{
  int c,res;

  static struct option options[]={{"output" ,required_argument,NULL,'o'}
                                 ,{"input"  ,required_argument,NULL,'i'}
                                 ,{"tables" ,no_argument      ,NULL,'t'}
                                 ,{"help"   ,no_argument      ,NULL,'h'}
                                 ,{"version",no_argument      ,NULL,'v'}
                                 ,{0,0,0,0}
                                 };

  res=0;

  while((c=getopt_long(argc,argv,"o:i:thv",options,NULL))!=-1)
  {
    switch(c)
    {
      case 'o' : *out=optarg;
                 break;
      case 'i' : *in=optarg;
                 break;
      case 't' : res+=1;
                 break;
      case 'h' : puts("rlp version 1.5, Copyright (C) 2006, 2009, 2015 Rui Carlos Goncalves\n"
                      "rlp comes with ABSOLUTELY NO WARRANTY.  This is free software, and\n"
                      "you are welcome to redistribute it under certain conditions.  See the GNU\n"
                      "General Public Licence for details.\n\n"
                      "Solve linear programming problems using simplex method.\n\n"
                      "Options:\n"
                      "  --help                 Print this help\n"
                      "  --version              Print version info\n"
                      "  --input=<file>         Input file (default: stdin)\n"
                      "  --tables               Print all tables\n"
                      "  --output=<file>        Output file (default: stdout)"
                     );
                 res-=2;
                 break;
      case 'v' : puts("rlp version 1.5, Copyright (C) 2006, 2009, 2015 Rui Carlos Goncalves\n"
                      "rlp comes with ABSOLUTELY NO WARRANTY.  This is free software, and\n"
                      "you are welcome to redistribute it under certain conditions.  See the GNU\n"
                      "General Public Licence for details."
                     );
                 res-=2;
                 break;
      default  : break;
    }
  }

  if(optind<argc){
    fputs(errors[4],stderr);
    exit(5);
  }

  return res;
}
